executor/new_distsql.go

```diff
@@ -286,62 +289,196 @@ type IndexLookUpExecutor struct {
+		e.ctx.GetSessionVars().DistSQLScanConcurrency, e.keepOrder, e.desc, getIsolationLevel(e.ctx.GetSessionVars()), e.priority)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	result.Fetch(e.ctx.GoCtx())
+	ih := &e.indexHandler
+	ih.wg.Add(1)
+	go func() {
+		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
+		ih.fetchHandles(e, result, workCh, ctx, finished)
+		cancel()
+		if err := result.Close(); err != nil {
+			log.Error("close SelectDAG result failed:", errors.ErrorStack(err))
+		}
+		close(workCh)
+		ih.wg.Done()
+	}()
+	return nil
+}
+
+// fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
+func (ih *indexHandler) fetchHandles(e *IndexLookUpExecutor, result distsql.SelectResult, workCh chan<- *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+	for {
+		handles, finish, err := extractHandlesFromIndexResult(result)
+		if err != nil {
+			workCh <- &lookupTableTask{
+				tasksErr: errors.Trace(err),
+			}
+			return
+		}
+		if finish {
+			return
+		}
+		tasks := e.buildTableTasks(handles)
+		for _, task := range tasks {
+			select {
+			case <-ctx.Done():
+				return
+			case <-finished:
+				return
+			case workCh <- task:
+			}
+		}
+	}
+}
+
+func (ih *indexHandler) close() {
+	ih.wg.Wait()
+}
+
+// tableHandler is used by IndexLookUpExecutor to maintain table lookup background goroutines.
+type tableHandler struct {
+	taskChan chan *lookupTableTask
+	taskCurr *lookupTableTask
+}
+
+// openTableHandler open the tableHandler, launch some background worker goroutines which pick
+// tasks from workCh, execute the task and store the results in tableHandler's taskChan.
+func (e *IndexLookUpExecutor) openTableHandler(workCh <-chan *lookupTableTask, finished <-chan struct{}) {
+	th := &e.tableHandler
+	th.taskChan = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
+	var wg sync.WaitGroup
+	wg.Add(lookupConcurrencyLimit)
+	for i := 0; i < lookupConcurrencyLimit; i++ {
+		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
+		go func() {
+			th.pickAndExecTask(e, workCh, ctx, finished)
+			cancel()
+			wg.Done()
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(th.taskChan)
+	}()
+}
+
+// pickAndExecTask picks tasks from workCh, execute them, and store the result to tableHandler's taskCh.
+func (th *tableHandler) pickAndExecTask(e *IndexLookUpExecutor, workCh <-chan *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+	for {
+		select {
+		case task, ok := <-workCh:
+			if !ok {
+				return
+			}
+			if task.tasksErr != nil {
+				th.taskChan <- task
+				return
+			}
+
+			// TODO: The results can be simplified when new_distsql.go replace distsql.go totally.
+			e.executeTask(task, ctx)
+			th.taskChan <- task
+		case <-ctx.Done():
+			return
+		case <-finished:
+			return
+		}
+	}
+}
+
+func (th *tableHandler) next() (Row, error) {
```

Comment:
I think move this method to IndexLookUpExecutor is better, it's a foreground function, tableHandler is supposed to handle background work.